File: apps/web/scripts/extract-presets.py
#!/usr/bin/env python3
import re
import sys
from pathlib import Path

root = Path(__file__).parent.parent
presets_md = root / 'docs' / 'PRESETS.md'
presets_ts = root / 'app' / 'config' / 'presets.ts'

with open(presets_md, 'r') as f:
    content = f.read()

# Fix imports
content = content.replace(
    "import { ThemePreset } from '../types/theme'",
    "import type { ThemePreset, ThemePresets } from '~/types/theme'"
)

# Fix type
content = content.replace('Record<string, ThemePreset>', 'ThemePresets')

# Add source property to presets missing it
# Match pattern: 'preset-name': { ... label:
def add_source_if_missing(match):
    full_match = match.group(0)
    # Check if source already exists in this preset block
    if 'source:' in full_match:
        return full_match
    # Add source before label
    return re.sub(r'(label:)', r"    source: 'BUILT_IN',\n    \1", full_match, count=1)

# Match preset blocks that start with 'name': { and have label: soon after
content = re.sub(
    r"('[\w-]+':\s*\{[^}]*?)(label:)",
    add_source_if_missing,
    content,
    flags=re.DOTALL
)

with open(presets_ts, 'w') as f:
    f.write(content)

# Count presets
preset_count = len(re.findall(r"'[\w-]+':\s*\{", content))
print(f"✅ Successfully extracted {preset_count} presets")
print(f"   File: {presets_ts}")

File: apps/web/scripts/extract-presets.js
const fs = require('fs')
const path = require('path')

const presetsMdPath = path.join(process.cwd(), 'docs', 'PRESETS.md')
const presetsTsPath = path.join(process.cwd(), 'app', 'config', 'presets.ts')

try {
  const content = fs.readFileSync(presetsMdPath, 'utf-8')

  // Find the start of the TypeScript code (after ```ts)
  const startMarker = '```ts'
  const endMarker = '```'

  const startIdx = content.indexOf(startMarker)
  if (startIdx === -1) {
    console.error('Could not find start marker')
    process.exit(1)
  }

  // Find the end marker after the start
  const endIdx = content.indexOf(endMarker, startIdx + startMarker.length)
  if (endIdx === -1) {
    console.error('Could not find end marker')
    process.exit(1)
  }

  // Extract the content between markers
  const tsCode = content.substring(startIdx + startMarker.length, endIdx).trim()

  // Update the import to match our structure
  const updatedCode = tsCode
    .replace(
      "import { ThemePreset } from '../types/theme'",
      "import type { ThemePreset, ThemePresets } from '~/types/theme'",
    )
    .replace('Record<string, ThemePreset>', 'ThemePresets')

  // Add source property to all presets
  const withSource = updatedCode.replace(/(\w+):\s*\{[\s\S]*?label:/g, (match) => {
    // Check if source already exists
    if (match.includes('source:')) {
      return match
    }
    // Add source after the opening brace
    return match.replace('{', "{\n    source: 'BUILT_IN',")
  })

  // Write to presets.ts
  fs.writeFileSync(presetsTsPath, withSource, 'utf-8')

  // Count presets
  const presetCount = (withSource.match(/'[^']+':\s*\{/g) || []).length
  console.log(`✅ Successfully extracted ${presetCount} presets from PRESETS.md to app/config/presets.ts`)
} catch (error) {
  console.error('Error extracting presets:', error)
  process.exit(1)
}
